### PR TITLE
feat: Basically do a whole re-write and add some basic git functionality

### PR DESCRIPTION
**Pull request by BenjaminMichaelis**
Contributes to #4

Basically an entire logic and architecture re-write

### CODE REVIEW

**Comment by BenjaminMichaelis** [comment]
@Keboo Also, if you have any idea why git move when looking at the file state sees the file as NewInIndex instead of RenamedInIndex (unit test and state verified through git gui and cli) but gitkraken shows as renamed, id be interested. Not sure if I'm doing git move incorrectly or similar, or if once it gets committed it will appear as a rename

File: ListingManager/ListingManager.cs [hunk 1]
```diff
@@ -24,15 +40,15 @@ public static IEnumerable<string> GetAllExtraListings(string pathToStartFrom)
         }
     }
 
-    private static bool TryGetListing(string listingPath, out ListingInformation? listingData)
+    private static bool TryGetListing(string listingPath, out ListingInformation? listingData, bool isTest = false)
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:43]:
```suggestion
    private static bool TryGetListing(string listingPath, [NotNullWhen(true)]out ListingInformation? listingData, bool isTest = false)
```

File: ListingManager/ListingManager.cs [hunk 2]
```diff
@@ -318,38 +204,79 @@ public static bool GetPathToAccompanyingUnitTest(string listingPath, out string
...
+        {
+            bool result = TryGetListing(fileName, out ListingInformation? data);
+            if (result)
             {
-                FileName = fileName,
-                Arguments = "-c \"" + command + "\"",
-                UseShellExecute = false,
-                RedirectStandardOutput = true,
-                CreateNoWindow = true
+                if (data is not null)
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:220]: Check not needed once annotation is updated.

File: ListingManager/ListingManager.cs [hunk 3]
```diff
@@ -318,38 +204,79 @@ public static bool GetPathToAccompanyingUnitTest(string listingPath, out string
...
+    public static List<ListingInformation> PopulateListingDataFromPath(string pathToChapter, bool singleDir)
+    {
+        List<ListingInformation> listingData = new();
+        List<ListingInformation> testListingData = new();
+        var listingFiles = FileManager.GetAllFilesAtPath(pathToChapter)
+            .OrderBy(x => x).ToList();
+        foreach (string fileName in listingFiles)
+        {
+            bool result = TryGetListing(fileName, out ListingInformation? data);
+            if (result)
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:218]:
```suggestion
            if (TryGetListing(fileName, out ListingInformation? data))

```

File: ListingManager/ListingManager.cs [hunk 4]
```diff
@@ -318,38 +204,79 @@ public static bool GetPathToAccompanyingUnitTest(string listingPath, out string
...
+                {
+                    throw new InvalidOperationException("Listing data is unexpectedly null with a successful result");
+                }
+            }
+        }
+
+        if (!singleDir)
+        {
+            var listingTestFiles = FileManager.GetAllFilesAtPath($"{pathToChapter}.Tests")
+    .OrderBy(x => x).ToList();
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:241]: do we need the `.ToList()` here or can we just let the foreach iterate over it?

File: ListingManager/ListingManager.cs [hunk 5]
```diff
@@ -318,38 +204,79 @@ public static bool GetPathToAccompanyingUnitTest(string listingPath, out string
...
+            foreach (string fileName in listingTestFiles)
+            {
+                bool result = TryGetListing(fileName, out ListingInformation? data, true);
+                if (result)
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:245]:
```suggestion
                if (TryGetListing(fileName, out ListingInformation? data, true))
```

File: ListingManager/ListingManager.cs (hunk 1 above)
Comment (Keboo) [inline ListingManager/ListingManager.cs:43]: It is a bit awkward to have an out parameter listed before an in parameter. I would suggest, just doing a method overload instead.

File: ListingManager/ListingManager.cs [hunk 6]
```diff
@@ -318,38 +204,79 @@ public static bool GetPathToAccompanyingUnitTest(string listingPath, out string
...
-        {
-            StartInfo = new ProcessStartInfo
+    }
+
+    public static List<ListingInformation> PopulateListingDataFromPath(string pathToChapter, bool singleDir)
+    {
+        List<ListingInformation> listingData = new();
+        List<ListingInformation> testListingData = new();
+        var listingFiles = FileManager.GetAllFilesAtPath(pathToChapter)
+            .OrderBy(x => x).ToList();
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:214]: Do we need the `.ToList()` here or can we let the foreach iterate?

File: ListingManager/ListingManager.cs [hunk 7]
```diff
@@ -318,38 +204,79 @@ public static bool GetPathToAccompanyingUnitTest(string listingPath, out string
...
+                        throw new InvalidOperationException("Listing data is unexpectedly null with a successful result");
+                    }
+                }
             }
-        };
+        }
+        foreach (ListingInformation testListingInformation in testListingData)
+        {
+            ListingInformation listingInformation = listingData.Where(x => x.OriginalListingNumber == testListingInformation.OriginalListingNumber && x.OriginalChapterNumber == testListingInformation.OriginalChapterNumber && x.OriginalListingNumberSuffix == testListingInformation.OriginalListingNumberSuffix).First();
+            if (testListingInformation.Caption.ToLowerInvariant() == "Tests".ToLowerInvariant() && listingInformation.Caption != string.Empty)
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:261]: You don't want to do case-insensitive comparisons like this. A better option is to use the overloads on `string.Equals` that let you specify that you don't care about case. Doing it this way allocates extra strings that are not needed.

File: ListingManager/ListingManager.cs [hunk 8]
```diff
@@ -318,38 +204,79 @@ public static bool GetPathToAccompanyingUnitTest(string listingPath, out string
...
+                    {
+                        throw new InvalidOperationException("Listing data is unexpectedly null with a successful result");
+                    }
+                }
             }
-        };
+        }
+        foreach (ListingInformation testListingInformation in testListingData)
+        {
+            ListingInformation listingInformation = listingData.Where(x => x.OriginalListingNumber == testListingInformation.OriginalListingNumber && x.OriginalChapterNumber == testListingInformation.OriginalChapterNumber && x.OriginalListingNumberSuffix == testListingInformation.OriginalListingNumberSuffix).First();
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:260]:
```suggestion
            ListingInformation listingInformation = listingData.First(x => x.OriginalListingNumber == testListingInformation.OriginalListingNumber && x.OriginalChapterNumber == testListingInformation.OriginalChapterNumber && x.OriginalListingNumberSuffix == testListingInformation.OriginalListingNumberSuffix);

```

File: ListingManager.Tests/FileManagerTests.cs [hunk 9]
```diff
@@ -1,16 +1,45 @@
...
+            "    using System;",
+            "    using System.Reflection;",
+            "    public class Program { }",
+            "}"
+        };
+
+        DirectoryInfo tempDir = CreateTempDirectory(new(Path.GetTempPath()));
+        CreateTempDirectory(tempDir, name: chapterFilePath);
+        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
+        Xunit.Assert.Single(writtenFiles);
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply ListingManager.Tests/FileManagerTests.cs:41]: Migrated to all Xunit

File: ListingManager.Tests/TempFileTestBase.cs [hunk 10]
```diff
@@ -92,6 +91,38 @@ private string GetPath(DirectoryInfo? parentDirectory, string? name, string? ext
...
+        {
+            fullPaths.Add(Path.Combine(targetDirectory?.FullName ?? TempDirectory.FullName, fileName));
+        }
+
+        return fullPaths;
+    }
+
+    public FileInfo WriteFile(DirectoryInfo targetDirectory, string fileName, List<string> toWrite)
+    {
+        var ret = CreateTempFile(targetDirectory, name: fileName, contents: toWrite.ToString());
```
[the remark this answers is not in the document]
Comment (BenjaminMichaelis) [reply ListingManager.Tests/TempFileTestBase.cs:109]: I want to re-write this method, this is one of the ones I just moved around to a class that made more sense, but haven't re-written completely yet, and plan on doing later in #88

File: ListingManager/ListingManager.cs (hunk 1 above)
Comment (BenjaminMichaelis) [reply ListingManager/ListingManager.cs:43]: Nice!

Comment (BenjaminMichaelis) [reply ListingManager/ListingManager.cs]: I agree, fixed

File: ListingManager.Tests/ListingInformationTests.cs [hunk 11]
```diff
@@ -1,45 +1,241 @@
...
+
+        ListingInformation listingInformation = new(writtenFiles.First().FullName);

+        Assert.NotNull(listingInformation);

+        Assert.Equal(Path.GetExtension(listing), listingInformation.ListingExtension);
+        Assert.False(listingInformation.Changed);
+    }

    [Theory]
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:85]:
```suggestion
    }
    
    [Theory]

```

File: ListingManager.Tests/ListingInformationTests.cs [hunk 12]
```diff
@@ -1,45 +1,241 @@
...
+            "    public class Program { }",
+            "}"
+        };
+
+        DirectoryInfo tempDir = CreateTempDirectory(new(Path.GetTempPath()));
+        WriteFiles(tempDir, filesToMake, toWrite);
+        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
+        Assert.Single(writtenFiles);
+
+        Assert.Throws<System.ArgumentException>(() => new ListingInformation(writtenFiles.First().FullName));
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:145]:
```suggestion
        var writtenFile = Assert.Single(writtenFiles);

        Assert.Throws<System.ArgumentException>(() => new ListingInformation(writtenFile.FullName));

```

File: ListingManager.Tests/ListingInformationTests.cs [hunk 13]
```diff
@@ -1,45 +1,241 @@
...
+            "namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter18.Listing18_01",
+            "{",
+            "    using System;",
+            "    using System.Reflection;",
+            "    public class Program { }",
+            "}"
+        };
+
+        DirectoryInfo tempDir = CreateTempDirectory(new(Path.GetTempPath()));
+        WriteFiles(tempDir, filesToMake, toWrite);
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:141]: This seems to duplicate the next line, do we need to call this twice?

File: ListingManager.Tests/ListingInformationTests.cs [hunk 14]
```diff
@@ -1,45 +1,241 @@
...
+        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:174]: Same as the prior comment, do we need to invoke WriteFiles twice?

File: ListingManager.Tests/ListingInformationTests.cs [hunk 15]
```diff
@@ -1,45 +1,241 @@
...
+        Assert.Single(writtenFiles);
+
+        Assert.Equal(expected, (new ListingInformation(writtenFiles.First().FullName)).GetPaddedListingNumberWithSuffix());
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:177]:
```suggestion
        var writtenFile = Assert.Single(writtenFiles);

        Assert.Equal(expected, (new ListingInformation(writtenFile.FullName)).GetPaddedListingNumberWithSuffix());

```

File: ListingManager.Tests/ListingInformationTests.cs [hunk 16]
```diff
@@ -1,45 +1,241 @@
...
+            "    public class Program { }",
+            "}"
+        };
+
+        DirectoryInfo tempDir = CreateTempDirectory(new(Path.GetTempPath()));
+        WriteFiles(tempDir, filesToMake, toWrite);
+        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
+        Assert.Single(writtenFiles);
+
+        Assert.Equal(expected, new ListingInformation(writtenFiles.First().FullName, isTest).GetNewNamespace(chapterOnly));
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:212]: Same as above tests

File: ListingManager.Tests/ListingInformationTests.cs [hunk 17]
```diff
@@ -1,45 +1,241 @@
...
+            "    public class Program { }",
+            "}"
+        };
+
+        DirectoryInfo tempDir = CreateTempDirectory(new(Path.GetTempPath()));
+        WriteFiles(tempDir, filesToMake, toWrite);
+        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
+        Assert.Single(writtenFiles);
+
+        Assert.Equal(expected, new ListingInformation(writtenFiles.First().FullName, isTest).GetNewFileName(chapterOnly));
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:239]: Same as above tests

File: ListingManager/IStorageManager.cs [hunk 18]
```diff
@@ -0,0 +1,6 @@
+﻿namespace EssentialCSharp.ListingManager;
+
+public interface IStorageManager
+{
+    public void Move(string oldPath, string newPath);
```
Comment (Keboo) [inline ListingManager/IStorageManager.cs:5]:
```suggestion
    void Move(string oldPath, string newPath);

```

File: ListingManager/ListingManager.cs [hunk 19]
```diff
@@ -50,212 +70,82 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
...
+
+            string newNamespace = curListingData.GetNewNamespace(ChapterOnly);
+            string newFileName = curListingData.GetNewFileName(ChapterOnly);
+
+            Console.WriteLine($"Corrective action. {Path.GetFileName(curListingData.Path)} rename to {newFileName}");
+
+            if (!preview) UpdateNamespaceOfPath(curListingData.Path, newNamespace, newFileName);
 
-            if (testListingData.Where(x => x?.ListingNumber == curListingData.ListingNumber && x.ListingSuffix == curListingData.ListingSuffix).FirstOrDefault() is ListingInformation curTestListingData)
+            if (listingData.Where(item => item.AssociatedTest is not null).Where(x => x?.OriginalListingNumber == curListingData.OriginalListingNumber && x.OriginalListingNumberSuffix == curListingData.OriginalListingNumberSuffix).FirstOrDefault() is ListingInformation curTestListingData)
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:101]:
```suggestion
            if (listingData.FirstOrDefault(item => item.AssociatedTest is not null).Where(x => x?.OriginalListingNumber == curListingData.OriginalListingNumber && x.OriginalListingNumberSuffix == curListingData.OriginalListingNumberSuffix) is ListingInformation curTestListingData)
```

File: ListingManager.Tests/ListingInformationTests.cs (hunk 11 above)
Comment (BenjaminMichaelis) [reply ListingManager.Tests/ListingInformationTests.cs:85]: Something weird is happening with these, you marked one of these before as well but it looked normal in VS

image

Comment (Keboo) [reply ListingManager.Tests/ListingInformationTests.cs:85]: Turn on witespace, you have a character there that is not a normal space.

File: ListingManager.Tests/ListingInformationTests.cs [hunk 20]
```diff
@@ -1,45 +1,238 @@
...
+            "{",
+            "    using System;",
+            "    using System.Reflection;",
+            "    public class Program { }",
+            "}"
+        };
+
+        DirectoryInfo tempDir = CreateTempDirectory(new(Path.GetTempPath()));
+        WriteFiles(tempDir, filesToMake, toWrite);
+        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:33]:
```suggestion
        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
```

File: ListingManager.Tests/ListingInformationTests.cs [hunk 21]
```diff
@@ -1,45 +1,238 @@
...
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:75]:
```suggestion
        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
```

File: ListingManager.Tests/ListingInformationTests.cs [hunk 22]
```diff
@@ -1,45 +1,238 @@
...
```
Comment (Keboo) [inline ListingManager.Tests/ListingInformationTests.cs:112]:
```suggestion
        var writtenFiles = WriteFiles(tempDir, filesToMake, toWrite);
```

**Review by Keboo — APPROVED** [review]

**Pull request merged**